libr/core/dmh_glibc.inc.c

```diff
@@ -99,6 +99,215 @@ static GHT GH(get_main_arena_with_symbol)(RCore *core, RDebugMap *map) {
+			break;
+		}
+	}
+
+	if (!found_section) {
+		R_LOG_WARN ("get_section_bytes: section %s not found", section_name);
+		goto cleanup_exit;
+	}
+
+	// eprintf ("get_section_bytes: section found: %s size: %#08x  paddr: %#08x\n", section_name, size, paddr);
+
+	buf = calloc (size, 1);
+	if (!buf) {
+		R_LOG_ERROR ("get_section_bytes: calloc failed");
+		goto cleanup_exit;
+	}
+
+	st64 read_size = r_buf_read_at (libc_bf->buf, paddr, buf, size);
+
+	if (read_size != size) {
+		R_LOG_ERROR ("get_section_bytes: section read unexpected size: %#08x  (section->size: %d)", read_size, size);
+		free (buf);
+		buf = NULL;
+	}
+
+cleanup_exit:
+	r_bin_file_delete (bin, libc_bf->id);
+	r_bin_file_set_cur_binfile (bin, bf);
+	return buf;
+}
+
+R_API double GH(get_glibc_version)(RCore *core, const char *libc_path) {
+	double version = 0.0;
+
+	// First see if there is a "__libc_version" symbol
+	// If yes read version from there
+	GHT version_symbol = GH (get_va_symbol) (core, libc_path, "__libc_version");
+	if (version_symbol != GHT_MAX) {
+		FILE *libc_file = fopen (libc_path, "r ");
+		if (libc_file == NULL) {
+			R_LOG_WARN ("resolve_glibc_version: Failed to open %s", libc_path);
+			return false;
+		}
+		// TODO: futureproof this
+		char version_buffer[5];
+		fseek (libc_file, version_symbol, SEEK_SET);
+		fread (version_buffer, 1, 4, libc_file);
+		fclose (libc_file);
+		if (!r_regex_match ("\\d.\\d\\d", "e", version_buffer)) {
+			R_LOG_WARN ("resolve_glibc_version: Unexpected version format: %s", version_buffer);
+			return false;
+		}
+		version = strtod (version_buffer, NULL);
+		R_LOG_INFO ("libc version %.2f identified from symbol", version);
+		return version;
+	}
+
+	// Next up we try to read version from banner in .rodata section
+	// also inspired by pwndbg
+	const ut8 *rodata_bytes = GH (get_section_bytes) (core, libc_path, ".rodata");
+	GHT rodata_size = GH (get_section_size) (core, libc_path, ".rodata");
+	const ut8 *banner_start = NULL;
+	if (rodata_bytes != NULL) {
+		banner_start = r_mem_mem (rodata_bytes, rodata_size, (const ut8 *)"GNU C Library", strlen ("GNU C Library"));
+	}
+
+	if (banner_start != NULL) {
+		// eprintf("banner found: %s\n", banner_start);
+		RRegex *rx = r_regex_new ("release version (\\d.\\d\\d)", "en");
+		RList *matches = r_regex_match_list (rx, (const char *)banner_start);
+		const char *version_start;
+		// We only care about the first match
+		const char *first_match = r_list_first (matches);
+		if (first_match)	{
+			version_start = first_match + strlen ("release version ");
+			version = strtod (version_start, NULL);
+		}
+		r_list_free (matches);
+		r_regex_free (rx);
+	}
+
+	if (version != 0) {
+		R_LOG_INFO ("libc version %.2f identified from .rodata banner", version);
+		return version;
+	}
+
+	R_LOG_WARN ("get_glibc_version failed");
+	return version;
+}
+
+static bool GH(resolve_glibc_version)(RCore *core) {
+	r_return_val_if_fail (core && core->dbg && core->dbg->maps, false);
+
+	double version = 0;
```

Comment:
im not a big fan of using double/float for version numbers

Reply:
both versions were in the code already like in old is_tcache, personally i think double reads better in code